### PR TITLE
Fixed #28382 -- Prevented BaseExpression._output_field from being set if _resolve_output_field() fails.

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://code.djangoproject.com/ticket/28382

### CODE REVIEW

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -253,18 +253,12 @@ def _resolve_output_field(self):
...
-            if num_sources == 0:
-                self._output_field = None
-            else:
-                for source in sources:
-                    if self._output_field is None:
-                        self._output_field = source
-                    if source is not None and not isinstance(self._output_field, source.__class__):
-                        raise FieldError(
-                            "Expression contains mixed types. You must set output_field")
+        sources_iter = filter(None.__ne__, self.get_source_fields())
```
Comment (charettes) [inline django/db/models/expressions.py]: Might want to use a generator here to make it more readable `(field for field in self.get_source_fields() if field is not None)`.

File: django/db/models/expressions.py [hunk 2]
```diff
@@ -253,18 +253,12 @@ def _resolve_output_field(self):
...
+        for source in sources_iter:
+            output_field = source
```
Comment (charettes) [inline django/db/models/expressions.py]: Iterate using `output_field` directly instead?

File: django/db/models/expressions.py (hunk 1 above)
Comment (sir-sigurd) [inline django/db/models/expressions.py]: I considered it, but it seemed too verbose too me.

Comment (timgraham) [inline django/db/models/expressions.py]: I'd also favor Simon's version. The current version is too cryptic, in my opinion. Also: `sources_iter` -> sources`?

Comment (sir-sigurd) [inline django/db/models/expressions.py]: IMO `sources_iter` is better because it highlights that it's iterator.

**Comment by sir-sigurd** [comment]
Created ticket https://code.djangoproject.com/ticket/28382.

**Pull request merged**